com.unity.perception/Runtime/GroundTruth/Labeling/LabelManager.cs

```diff
@@ -0,0 +1,167 @@
+﻿using System.Collections.Generic;
+using UnityEngine.Perception.Randomization.Randomizers;
+
+namespace UnityEngine.Perception.GroundTruth
+{
+    /// <summary>
+    /// Manages the registration of <see cref="Labeling"/> components
+    /// </summary>
+    public class LabelManager
+    {
+        /// <summary>
+        /// Returns the active LabeledObjectsManager instance
+        /// </summary>
+        public static LabelManager singleton { get; } = new LabelManager();
+
+        const uint k_StartingIndex = 1;
+        uint m_CurrentObjectIndex = k_StartingIndex;
+        List<IGroundTruthGenerator> m_ActiveGenerators = new List<IGroundTruthGenerator>();
+        LinkedHashSet<Labeling> m_LabelsPendingRegistration = new LinkedHashSet<Labeling>();
```

Comment:
LinkedHashSet is clever, but it allocates GC memory on every insertion due to the linked list. We should be able to do better than this. I think this is worth fixing before landing the PR because otherwise it adds to our growing GC tech debt.

Reply:
I'm open to suggestions, but LinkedHashSet is great in that it provides O(1) lookup, insertion, and removal. The LabelManager will require the capability to remove arbitrary registered LabelingComponents from it's registry, so the only alternative data structure I can think of is a normal array in which removal would be an O(n^2) operation.

We've haven't seen any performance issues when employing this data structure with the TagManager either. The use of the AddComponentObject API in the previous Labeling and GroundTruthLabelSetupSystem implementation nullified any real-benefit from using ECS. Here is a quote from the [Unity ECS documentation](https://docs.unity3d.com/Packages/com.unity.entities@0.0/api/Unity.Entities.EntityManager.html) to clarify:
> Accessing data in a managed object forfeits many opportunities for increased performance. Adding managed objects to an entity should be avoided or used sparingly. Adding a component changes an entity's archetype and results in the entity being moved to a different chunk.

I think implementing a more complex alternative Label registration data structure would be preemptively optimizing a solution that has no data supporting that it is under-performing. Actually, I would be inclined to believe that the new LinkedHashSet implementation is more performant than the previous ECS implementation.

Reply:
Linked lists provide no cache coherence and extra GC allocations, so it shouldn't take a deep dive to see that this is an inefficient data structure. I am not going to fight on this, since I realize we have some much bigger performance issues, but I don't think it is wise to treat poorly performing code as a "will not cause issues" kind of problem, since poor performance is generally a death by a thousand cuts kind of problem.

One thing that would make this structure more efficient would be to swap out the LinkedList<T> for a LinkedList<List<T>>, where you allocate in chunks to amortize the GC allocations and memory jumping while still maintaining O(1) on other operations.

Reply:
My point with the AddComponentObject API is to prove that we didn't necessarily have cache coherence in the first place (new chunk allocation required per new entity for SharedComponents), so the new LinkedHashSet solution is likely equivalent or faster than our previous implementation due to reduced ECS overhead. In this sense, the new data structure should have the same optimization prioritization as the former ECS solution (which wasn't even on our radar IIRC).

In regards to your suggestion, do you have a link to an existing implementation of this data structure I could reference? It sounds very neat.

Reply:
I hear you - that's a good point. Here is an idea for your consideration: https://www.jacksondunstan.com/articles/4963
There are lots of ways to improve this system honestly, but I agree we need to take some extra time to come back to performance as a whole on this type of project.